### PR TITLE
docs: Update contributing docs for release workflow

### PR DESCRIPTION
**Pull request by zioroboco**
https://github.com/cultureamp/kaizen-design-system/blob/louis/release-docs/CONTRIBUTING.md#releasing-packages

This document is getting a bit meandering, so I think we should consider splitting these topics into the wiki on the repo at some point, and linking to them there from the README etc.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -61,31 +61,91 @@ You can also clean up generated files:
 
 `yarn clean`
 
+## Releasing packages
 
-## Pull Requests, versioning and releasing
+Automated releases to the npm public registry are triggered for all pull requests containing modifications to one or more npm packages (found in the `/packages/` directory). The information required to determine the version update for each release is taken from the title and content of the pull request.
```
Comment (didoesdigital) [inline CONTRIBUTING.md]: Nice overview 👍

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -61,31 +61,91 @@ You can also clean up generated files:
...
+- If the branch contains multiple commits, the pull request must have a [conventional title](#conventional-commit)
+- If the branch contains only a single commit...
+  - ... the commit itself must have a [conventional commit message](#conventional-commit), and...
+  - ... (for consistency) the PR title must match that commit message
+
+Once that pull request is merged into master, an automated release will be triggered, and the newly published package version will be available on the npm public registry.
+
+### Semantic Versioning
+
+All npm packages follow strict semantic versioning (or _semver_). Semantic versioning gives specific meaning to changes in version numbers, as follows:
```
Comment (didoesdigital) [inline CONTRIBUTING.md]: Great clarification of where the behaviour comes from (all npm packages follow semver) 👍

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -61,31 +61,91 @@ You can also clean up generated files:
...
+> - `PATCH` version when you make backwards compatible bug fixes.
 
 — <https://semver.org/>
 
-If you make any breaking changes, include "BREAKING CHANGE" in the commit message. For example: **BREAKING CHANGE: removed feature from component X**.
+**Note that we do not update package version numbers directly**, but instead depend on a [Conventional Commit](#conventional-commit) workflow which will version and release packages according to the content of our pull requests.
+
+### Conventional Commit
+
+In order to provide versioning information to our automated release workflow, we ensure that our pull requests are structured in a way that the CI pipeline is able to infer the required version update (as well as which packages are to be released with that update) from the content of our pull requests.
```
Comment (didoesdigital) [inline CONTRIBUTING.md]: This sentence was hard to parse. Can we break it up?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -61,31 +61,91 @@ You can also clean up generated files:
...
+
+These constraints require that each pull request have a title formatted as follows:
+
+```
+<type>: <description>
+```
+
+#### Description
+
+The `description` should be a short (less than ~60 characters) summary of changes introduced in the release. This summary will be included in the CHANGELOG (see the [releases page](https://github.com/cultureamp/kaizen-design-system/releases)), along with the commit type, a link to the pull request which triggered the release, and a detailed summary of any breaking changes.
```
Comment (didoesdigital) [inline CONTRIBUTING.md]: This is a good inclusion, explaining the implication of how the description is used in the CHANGELOG.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -61,31 +61,91 @@ You can also clean up generated files:
...
 
-For another example, a PR title of `feat: add a new Toggle component` will create a **minor** release, bumping the version from 16.0.0 to 16.1.0.
+There are a number of possible commit types (e.g. `chore`, `docs`, `refactor` etc. — for a complete list, refer to the [Angular convention](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type) docs), but two have special meaning for the purpose of versioning package releases:
 
-Finally, a PR title of `fix: fix the focus styling on our button component` will create a **patch** release, bumping the version from 16.0.0 to 16.0.1.
+- `fix` — indicates that a change addressed a bug or security concern in the published code, but otherwise had no consequence for the released package's features or API (corresponding to a PATCH version)
 
-## Using your newly released component changes
+- `feat` — indicates that a change added something new to the released package's API or features without affecting existing functionality (corresponding to a MINOR version)
```
Comment (didoesdigital) [inline CONTRIBUTING.md]: Can we include a concrete example of `docs`? That's come up a few times and its absence here is noticed. That is, people look for how to update `docs` and it's not listed.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -16,7 +16,7 @@ To update a component:
 - Notify the front-end engineering practice (#pract_front_end_eng) of any possible breaking changes
 - Notify the QA practice of any possible breaking changes
 
-\* **If you're new to Kaizen, please ask the Design Systems Team (#team_design_systems) to set up an onboarding session to get you up to speed.** If you have an urgent PR to merge before that happens, it is safest to ask Design Systems Team to review it to catch any issues.
+**If you're new to Kaizen, please ask the Design Systems Team (#team_design_systems) to set up an onboarding session to get you up to speed.** If you have an urgent PR to merge before that happens, it is safest to ask Design Systems Team to review it to catch any issues.
```
Comment (sebpearce) [inline CONTRIBUTING.md]: The `\*` was intentional – there's a * in one of the bullet points above this bit.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -61,31 +61,91 @@ You can also clean up generated files:
...
-> Given a version number MAJOR.MINOR.PATCH, increment the:
-    MAJOR version when you make incompatible API changes,
-    MINOR version when you add functionality in a backwards compatible manner, and
-    PATCH version when you make backwards compatible bug fixes.
+To release a new package version, create a pull request which **modifies ONLY the package you wish to release**, and satisfies **ONE** of the following:
+
+- If the branch contains multiple commits, the pull request must have a [conventional title](#conventional-commit)
+- If the branch contains only a single commit...
+  - ... the commit itself must have a [conventional commit message](#conventional-commit), and...
+  - ... (for consistency) the PR title must match that commit message
```
Comment (sebpearce) [inline CONTRIBUTING.md]: Wonder if this could be simplified to something like:
```
To release a new package version, create a pull request which **modifies ONLY the package you wish to release**.

- Make sure your PR has a [conventional title](#conventional-commit)
- If there is only **one commit** in your PR, make sure that commit message matches the PR title
```
Will edit.

File: CONTRIBUTING.md (hunk 3 above)
Comment (sebpearce) [reply CONTRIBUTING.md]: Will have a crack at this.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -61,31 +61,91 @@ You can also clean up generated files:
...
 
-## Using your newly released component changes
+- `feat` — indicates that a change added something new to the released package's API or features without affecting existing functionality (corresponding to a MINOR version)
 
-To see your deployed Kaizen Component Library changes in your front-end codebase, you'll need to bump the version of the Kaizen package used. You might edit the `package.json` file to change the package version or run `yarn upgrade --latest @cultureamp/kaizen-component-library`.
+#### Breaking changes
+
+If a pull request includes changes which modify existing behaviour or APIs in a way that is not backwards compatible, that change needs to be marked with a `BREAKING CHANGE: <description>` line (including a description [as above](#description)) somewhere in the commit body of the merged commit to master, in order trigger a MAJOR version update on this release.
+
+Since we are using a squash-and-merge strategy for our pull requests, it is therefore recommended that breaking changes be introduced in their own commits, each with a commit summary in the format `BREAKING CHANGE: <description>`, and any further information written into the commit body.
```
Comment (sebpearce) [inline CONTRIBUTING.md]:
```suggestion
Since we are using a squash-and-merge strategy for our pull requests, we recommend that breaking changes be introduced in their own commits, each with a commit summary in the format `BREAKING CHANGE: <description>`, and any further information written into the commit body.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -61,31 +61,94 @@ You can also clean up generated files:
...
+#### Breaking changes
+
+In addition to `feat` and `fix` releases, if a pull request includes changes which modify existing behaviour or APIs in a way that is **not backwards compatible**, that change needs to be marked with a `BREAKING CHANGE: <description>` line (including a description [as above](#description)) somewhere in the commit body of the merged commit to master. Doing so will trigger a MAJOR version update in the corresponding release, irrespective of the commit type.
+
+Since we are using a squash-and-merge strategy for our pull requests, we recommend that you introduce breaking changes in their own commits, each with a commit summary in the format `BREAKING CHANGE: <description>`, with any additional notes in the commit body.
+
+> **Note:** Pull requests for branches containing a single commit are a special case, and should contain a commit with a conventional commit message (and a matching pull request title), with any `BREAKING CHANGE` annotations included in the commit body. To avoid this edge case, you can push an additional commit to your branch!
```
Comment (jasononeil) [inline CONTRIBUTING.md]: Does the `moar` app protect against this weirdness? Or will it still let me merge even though it'll result in a squashed commit without the desired effect?

Also - is it worth mentioning that you could push an empty commit? (Can you?)

Comment (zioroboco) [reply CONTRIBUTING.md]: You can push an empty commit... but it feels like a weird thing to bless as part of the workflow!

And yes, these constraints are all enforced by the [moai-merge](https://github.com/apps/moai-merge/) app.

Comment (zioroboco) [reply CONTRIBUTING.md]: To be clear, the app **does not** check that `BREAKING CHANGE: ...` annotations are included as commit message summaries. But it's a good idea, and doable!
